### ISSUE TITLE
[UML] Support Communication Diagram

### ISSUE THREAD

**Issue by xperiandri** [mermaid-js/mermaid#2647]
**Is your feature request related to a problem? Please describe.**
Support UML Communication Diagram

**Describe the solution you'd like**
New diagram type, UML Communication

**Comment by dingjimmy** [comment]
I support this.

I often find there are times when you wish to emphasise the relationships between components when diagramming the order of events in a system. Communication diagrams are great for this. While they share many core aspects with a sequence diagrams, the timeline orientated view of a sequence diagram sometimes just gets in way.

It would be incredibly useful for my and my co-workers to be able to author UML Communications diagrams using Mermaid.

**Comment by AndyBrownlie** [comment]
Yep, completely concur with this request 👍 . If I had any knowledge of how MermaidJS worked or my JS was better I'd offer a PR.

**Comment by jgreywolf** [comment]
Please provide an example of proposed syntax, and an image of the expected result

**Comment by au-phiware** [comment]
I propose a syntax that attempts to be in keeping with the existing sequence diagram and also follows the [current UML specification](https://www.omg.org/spec/UML/2.5.1/About-UML).  The following diagram (figure 17.26) and description can be found in [section 17.9.2](https://www.omg.org/spec/UML/2.5.1/PDF#%5B%7B%22num%22%3A2508%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C114.3%2C583.8%2C0%5D).

![Screenshot 2023-05-11 at 14-08-20 Unified Modeling Language v2 5 1 - formal-17-12-05 pdf](https://github.com/mermaid-js/mermaid/assets/816686/b94086bd-6ea1-4c61-880b-b230cc84469d)
> The Interaction described by a Communication Diagram in Figure 17.26 shows messages m1 and m3 being sent
concurrently from :r towards two instances of the part s. The sequence numbers show how the other messages are
sequenced. 1b.1 follows after 1b and 1b.1.1 thereafter etc. 2 follows after 1a and 1b.

```
communicationDiagram
    interaction M
    participant r as :r
    participant sk as s[k]:B
    participant su as s[u]:B

    1a:       r  --> sk: m1
    1b:       r  --> su: m3
    1b.1:     su --> sk: m3
    1b.1.1:   sk --> su: m3
    1b.1.1.1: sk --> su: m2
    2:        sk --> su: m2
```

The proposed syntax above introduces the interaction element which specifies the name and parameters of the [interaction](https://www.omg.org/spec/UML/2.5.1/PDF#%5B%7B%22num%22%3A2342%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C121.5%2C359.7%2C0%5D) and is rendered in the upper left pentagon of the frame. The lifelines are specified by participants and share the same semantics as the sequenceDiagram syntax. Messages are specified by a [sequence expression](https://www.omg.org/spec/UML/2.5.1/PDF#%5B%7B%22num%22%3A2501%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C121.5%2C400.3%2C0%5D), pair of participant, and message name, separated by `:`. The participants are separated by an arrow showing the message direction. Messages are decorated by a small arrow in the direction of the message close to the message name and sequence expression along the line between the participants. Whitespace is optional.

**Comment by au-phiware** [comment]
I made a start adapting sequenceDiagram.jison, I would change the signal rule to read:

```
signal
	: SEQ actor ARROW actor text2
	{ $$ = [$2,$4,{type: 'addMessage', seq:$1, from:$2.actor, to:$4.actor, msg:$5}] }
	;
```
Where SEQ is defined as:
```
[0-9]+[a-zA-Z]*(\.\s*[0-9]+[a-zA-Z]*)*\.?(?=[ \n]+)             { yytext = yytext.replace(/\s/g, ''); return 'SEQ'; }
```

(Looking at sequenceRenderer.ts, and feeling a little lost...)

**Comment by vjache** [comment]
This type of diagram should have been implemented in the first place. It is something that programmers drawing on the whiteboard when only even starting to consider architecture. The sequence diagrams are just instantioations/examples of what coud happen according to communication(also known 'collaboration') diagrams. It's a shame this type of diagram was not implemented in the first place.

**Comment by ExeVirus** [comment]
Communication diagrams and sequence diagrams convey very much the same data, but display it very differently:

[Example sequence vs communcation](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/what-is-communication-diagram/)

In the case of mermaid, current sequence diagrams can be easily converted to communication diagrams, assuming we add a way to annotate participant locations. I.e. a row, column (x,y) specification.

So to convert a sequence diagram to a communication diagram and have all the information required:

```
sequenceDiagram
    participant Alice
    participant Bob
    Bob->>Alice: Hi Alice
    Alice->>Bob: Hi Bob
```

```
communicationDiagram
    participant Alice 1 1
    participant Bob 2 1
    Bob->>Alice: Hi Alice
    Alice->>Bob: Hi Bob
```

where the syntax extension could simply be:

`participant <name> <row> <column>`

**Comment by ExeVirus** [comment]
I'm interested, yes. 

But unlikely to tackle it for at least several months, just had the time to put down the thought here in case someone else beats me to it.
